File: backend/app/api/routes/sub_account.py
"""REST endpoints for Binance Sub-Account data."""

from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field

from app.auth.user_auth import current_user_dep
from app.binance.sub_account_service import (
    SubAccountService,
    get_sub_account_service,
    is_sub_account_configured,
)
from app.core.exceptions import BinanceAPIError

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/sub-account", tags=["sub-account"])


class SubAccountConfigResponse(BaseModel):
    configured: bool


class SubAccountErrorResponse(BaseModel):
    error: str


def _get_service() -> SubAccountService:
    if not is_sub_account_configured():
        raise BinanceAPIError(
            "Sub-account API is not configured. "
            "Set BINANCE_SUB_ACCOUNT_API_KEY and BINANCE_SUB_ACCOUNT_API_SECRET."
        )
    return get_sub_account_service()


# ── Config ───────────────────────────────────────────────────────────


@router.get("/config", response_model=SubAccountConfigResponse)
async def sub_account_config(
    _user: Any = Depends(current_user_dep),
) -> SubAccountConfigResponse:
    return SubAccountConfigResponse(configured=is_sub_account_configured())


# ── Account ──────────────────────────────────────────────────────────


@router.get("")
async def get_sub_account_list(
    email: str | None = Query(default=None),
    is_freeze: str | None = Query(default=None),
    page: int | None = Query(default=None, ge=1),
    limit: int | None = Query(default=None, ge=1, le=500),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    accounts = await svc.get_sub_account_list(
        email=email,
        is_freeze=is_freeze,
        page=page,
        limit=limit,
    )
    return [a.model_dump() for a in accounts]


@router.get("/status")
async def get_sub_accounts_status(
    email: str | None = Query(default=None),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    statuses = await svc.get_sub_accounts_status(email=email)
    return [s.model_dump() for s in statuses]


@router.get("/assets")
async def get_sub_account_assets(
    email: str = Query(..., description="Sub-account email"),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    result = await svc.get_sub_account_assets(email=email)
    return result.model_dump()


@router.get("/spot-summary")
async def get_sub_account_spot_summary(
    email: str | None = Query(default=None),
    page: int | None = Query(default=None, ge=1),
    size: int | None = Query(default=None, ge=1, le=500),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    result = await svc.get_sub_account_spot_assets_summary(
        email=email,
        page=page,
        size=size,
    )
    return result.model_dump()


@router.get("/transaction-statistics")
async def get_sub_account_transaction_statistics(
    email: str | None = Query(default=None),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    return await svc.get_sub_account_transaction_statistics(email=email)


# ── Futures ──────────────────────────────────────────────────────────


@router.get("/futures")
async def get_futures_account(
    email: str = Query(..., description="Sub-account email"),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    result = await svc.get_futures_account(email=email)
    return result.model_dump()


@router.get("/futures/positions")
async def get_futures_position_risk(
    email: str = Query(..., description="Sub-account email"),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    result = await svc.get_futures_position_risk(email=email)
    return result.model_dump()


@router.get("/futures/summary")
async def get_futures_account_summary(
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=10, ge=1, le=500),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    result = await svc.get_futures_account_summary(page=page, limit=limit)
    return result.model_dump()


# ── Margin ───────────────────────────────────────────────────────────


@router.get("/margin")
async def get_margin_account(
    email: str = Query(..., description="Sub-account email"),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    result = await svc.get_margin_account(email=email)
    return result.model_dump()


@router.get("/margin/summary")
async def get_margin_account_summary(
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    result = await svc.get_margin_account_summary()
    return result.model_dump()


# ── Transfer history ─────────────────────────────────────────────────


@router.get("/transfers/spot")
async def get_spot_transfer_history(
    from_email: str | None = Query(default=None),
    to_email: str | None = Query(default=None),
    start_time: int | None = Query(default=None),
    end_time: int | None = Query(default=None),
    page: int | None = Query(default=None, ge=1),
    limit: int | None = Query(default=None, ge=1, le=500),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    transfers = await svc.get_spot_transfer_history(
        from_email=from_email,
        to_email=to_email,
        start_time=start_time,
        end_time=end_time,
        page=page,
        limit=limit,
    )
    return [t.model_dump() for t in transfers]


@router.get("/transfers/futures")
async def get_futures_transfer_history(
    email: str = Query(..., description="Sub-account email"),
    futures_type: int = Query(default=1, ge=1, le=2),
    start_time: int | None = Query(default=None),
    end_time: int | None = Query(default=None),
    page: int | None = Query(default=None, ge=1),
    limit: int | None = Query(default=None, ge=1, le=500),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    transfers = await svc.get_futures_transfer_history(
        email=email,
        futures_type=futures_type,
        start_time=start_time,
        end_time=end_time,
        page=page,
        limit=limit,
    )
    return [t.model_dump() for t in transfers]


@router.get("/transfers/universal")
async def get_universal_transfer_history(
    from_email: str | None = Query(default=None),
    to_email: str | None = Query(default=None),
    start_time: int | None = Query(default=None),
    end_time: int | None = Query(default=None),
    page: int | None = Query(default=None, ge=1),
    limit: int | None = Query(default=None, ge=1, le=500),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    transfers = await svc.get_universal_transfer_history(
        from_email=from_email,
        to_email=to_email,
        start_time=start_time,
        end_time=end_time,
        page=page,
        limit=limit,
    )
    return [t.model_dump() for t in transfers]


# ── Deposits ─────────────────────────────────────────────────────────


@router.get("/deposits")
async def get_deposit_history(
    email: str = Query(..., description="Sub-account email"),
    coin: str | None = Query(default=None),
    start_time: int | None = Query(default=None),
    end_time: int | None = Query(default=None),
    limit: int | None = Query(default=None, ge=1, le=500),
    _user: Any = Depends(current_user_dep),
) -> Any:
    svc = _get_service()
    result = await svc.get_deposit_history(
        email=email,
        coin=coin,
        start_time=start_time,
        end_time=end_time,
        limit=limit,
    )
    return result.model_dump()

File: frontend/src/api/subAccount.ts
import { apiFetch } from "./client"
import type {
  SubAccountInfo,
  SubAccountStatus,
  SubAccountAssetsResponse,
  SubAccountSpotSummary,
  FuturesAccountResponse,
  FuturesPositionRiskResponse,
  FuturesAccountSummaryResponse,
  MarginAccountResponse,
  MarginAccountSummaryResponse,
  SpotTransfer,
  FuturesTransfer,
  UniversalTransfer,
  DepositHistoryResponse,
  SubAccountConfigResponse,
} from "@/types/api"

export async function getSubAccountConfig(): Promise<SubAccountConfigResponse> {
  return apiFetch<SubAccountConfigResponse>("/sub-account/config")
}

export async function getSubAccounts(
  params?: { email?: string; page?: number; limit?: number }
): Promise<SubAccountInfo[]> {
  const query = new URLSearchParams()
  if (params?.email) query.set("email", params.email)
  if (params?.page) query.set("page", String(params.page))
  if (params?.limit) query.set("limit", String(params.limit))
  const qs = query.toString()
  return apiFetch<SubAccountInfo[]>(`/sub-account${qs ? `?${qs}` : ""}`)
}

export async function getSubAccountStatus(
  email?: string
): Promise<SubAccountStatus[]> {
  const query = email ? `?email=${encodeURIComponent(email)}` : ""
  return apiFetch<SubAccountStatus[]>(`/sub-account/status${query}`)
}

export async function getSubAccountAssets(
  email: string
): Promise<SubAccountAssetsResponse> {
  return apiFetch<SubAccountAssetsResponse>(
    `/sub-account/assets?email=${encodeURIComponent(email)}`
  )
}

export async function getSubAccountSpotSummary(
  email?: string
): Promise<SubAccountSpotSummary> {
  const query = email ? `?email=${encodeURIComponent(email)}` : ""
  return apiFetch<SubAccountSpotSummary>(
    `/sub-account/spot-summary${query}`
  )
}

export async function getFuturesAccount(
  email: string
): Promise<FuturesAccountResponse> {
  return apiFetch<FuturesAccountResponse>(
    `/sub-account/futures?email=${encodeURIComponent(email)}`
  )
}

export async function getFuturesPositions(
  email: string
): Promise<FuturesPositionRiskResponse> {
  return apiFetch<FuturesPositionRiskResponse>(
    `/sub-account/futures/positions?email=${encodeURIComponent(email)}`
  )
}

export async function getFuturesSummary(
  page = 1,
  limit = 10
): Promise<FuturesAccountSummaryResponse> {
  return apiFetch<FuturesAccountSummaryResponse>(
    `/sub-account/futures/summary?page=${page}&limit=${limit}`
  )
}

export async function getMarginAccount(
  email: string
): Promise<MarginAccountResponse> {
  return apiFetch<MarginAccountResponse>(
    `/sub-account/margin?email=${encodeURIComponent(email)}`
  )
}

export async function getMarginSummary(): Promise<MarginAccountSummaryResponse> {
  return apiFetch<MarginAccountSummaryResponse>(
    `/sub-account/margin/summary`
  )
}

export async function getSpotTransfers(params?: {
  from_email?: string
  to_email?: string
  start_time?: number
  end_time?: number
  page?: number
  limit?: number
}): Promise<SpotTransfer[]> {
  const query = new URLSearchParams()
  if (params?.from_email) query.set("from_email", params.from_email)
  if (params?.to_email) query.set("to_email", params.to_email)
  if (params?.start_time) query.set("start_time", String(params.start_time))
  if (params?.end_time) query.set("end_time", String(params.end_time))
  if (params?.page) query.set("page", String(params.page))
  if (params?.limit) query.set("limit", String(params.limit))
  const qs = query.toString()
  return apiFetch<SpotTransfer[]>(
    `/sub-account/transfers/spot${qs ? `?${qs}` : ""}`
  )
}

export async function getFuturesTransfers(params?: {
  email?: string
  futures_type?: number
  start_time?: number
  end_time?: number
  page?: number
  limit?: number
}): Promise<FuturesTransfer[]> {
  const query = new URLSearchParams()
  if (params?.email) query.set("email", params.email)
  if (params?.futures_type) query.set("futures_type", String(params.futures_type))
  if (params?.start_time) query.set("start_time", String(params.start_time))
  if (params?.end_time) query.set("end_time", String(params.end_time))
  if (params?.page) query.set("page", String(params.page))
  if (params?.limit) query.set("limit", String(params.limit))
  const qs = query.toString()
  return apiFetch<FuturesTransfer[]>(
    `/sub-account/transfers/futures${qs ? `?${qs}` : ""}`
  )
}

export async function getUniversalTransfers(params?: {
  from_email?: string
  to_email?: string
  start_time?: number
  end_time?: number
  page?: number
  limit?: number
}): Promise<UniversalTransfer[]> {
  const query = new URLSearchParams()
  if (params?.from_email) query.set("from_email", params.from_email)
  if (params?.to_email) query.set("to_email", params.to_email)
  if (params?.start_time) query.set("start_time", String(params.start_time))
  if (params?.end_time) query.set("end_time", String(params.end_time))
  if (params?.page) query.set("page", String(params.page))
  if (params?.limit) query.set("limit", String(params.limit))
  const qs = query.toString()
  return apiFetch<UniversalTransfer[]>(
    `/sub-account/transfers/universal${qs ? `?${qs}` : ""}`
  )
}

export async function getDepositHistory(params: {
  email: string
  coin?: string
  start_time?: number
  end_time?: number
  limit?: number
}): Promise<DepositHistoryResponse> {
  const query = new URLSearchParams({ email: params.email })
  if (params.coin) query.set("coin", params.coin)
  if (params.start_time) query.set("start_time", String(params.start_time))
  if (params.end_time) query.set("end_time", String(params.end_time))
  if (params.limit) query.set("limit", String(params.limit))
  return apiFetch<DepositHistoryResponse>(
    `/sub-account/deposits?${query.toString()}`
  )
}
